### DISCUSSION TITLE
System.Runtime.CompilerServices.AsynUnsafeAttribute - .NET7 Security

### DISCUSSION THREAD

**Discussion by c-ohle** [dotnet/runtime#72710] [Ideas] [answered]
### For security reasons and to avoid lots of code for security checks, it would be helpful to have an `Attribute` like `[System.Runtime.CompilerServices.AsynUnsafeAttribute]`.

This attribute should be applicable to class and value types or specific functions and properties.
The compiler should then simply forbid using the functions in asynchronous code.

Many classes and functions use `ThreadStatic` content and using it in an asynchronous context makes no sense anyway.
Doesn't make any sense like for example calling `Windows.Forms` functions directly, knowing that it leads to an `Exception`.

For .NET7 developments, in [discussion with other dotnet-devs](https://github.com/dotnet/runtime/issues/71791), we have to make core solutions absolutely end user save. 
In our case, for example, this requires millions of relatively slow `AsyncLocal<T>` requests at runtime - just to catch a possible programming error - just to raise a runtime `Exception`.
Not only does it slow the computations down, it also requires a lot of extra code and additional static root objects.

How it could look like:
```C#
[System.Runtime.CompilerServices.AsynUnsafeAttribute]
static class AClass 
{
  public static void AMethod();
}

async void AAsynFunc()
{
  AClass.AMethod(); // <- Error: CSXXXX "Not allowed in an asynchronous context."
  await SomeThing();
 
}
```
### This could help make C# more secure and faster.

**Comment by teo-tsirpanis** [comment, marked as answer]
Addressed in #72711.

**Reply by c-ohle** [reply]
Hi, I need to say thanks, I implemented this already and it works great with the ref struct wrapper. 
No side effects, no problems with debug, minimal overhead, no performance lost - things I espected.
And it's already tested in all possible configurations for NET6 / 7 and VS versions.
[If you are interested - as code](https://github.com/c-ohle/RationalNumerics/blob/master/System.Numerics.Rational/BigRational.Safe.cs)
[or documented](https://c-ohle.github.io/RationalNumerics/api/System.Numerics.html)
